### ISSUE TITLE
`Marker` antialiasing is broken

### ISSUE THREAD

**Issue by brisvag** [vispy/vispy#2208]
I'm not sure whether this is a consequence of general transparency issues or if it is in the shader code; try this:

```python
import numpy as np
import vispy.scene
from vispy.scene import visuals

canvas = vispy.scene.SceneCanvas(keys='interactive', show=True)
view = canvas.central_widget.add_view()

np.random.seed(1)
pos = np.random.normal(size=(100, 3))

scatter = visuals.Markers()
scatter.antialias = 10
scatter.set_data(pos, edge_color='blue', face_color='white', size=0.5, edge_width=10, scaling=True)

view.add(scatter)

view.camera = 'turntable'
view.camera.fov = 0

vispy.app.run()
```

I set an exaggerated `antialias` value to make it more obvious. It looks like a depth issue (since the stuff that's visible behind does not appear all in the right order). Any ideas?

**Comment by djhoese** [comment]
Just for reference, here is what it looks like (at least for me):

![image](https://user-images.githubusercontent.com/1828519/131687658-2e791ef7-475f-40a5-b1d9-38f27e06c3e8.png)

I feel like there are so many small things that could be at play here. In the last couple years things have changed related to rotation (fixed by @sofroniewn), depth testing was turned on in 2016, and changes to using discard versus just `if` statements. Or this could just be another bug or a limitation of OpenGL when two things are drawn at the same "depth".

**Comment by brisvag** [comment]
I tried to manually set `gl_FragDepth`, and that seems to do nothing different. I don't think that's the problem, because the opaque areas are rendered at the right place.

Actually, this prompted me to simply try to set a lower alpha for the whole marker. Indeed, this does not seem to be just an antialiasing issue, but an alpha issue in general. Maybe will be addressed by #2039? I wonder if @asnt could help us here?

Here's another code snippet, which displays this:
![image](https://user-images.githubusercontent.com/23482191/131690864-f8c60e70-0c4f-48ef-af7a-3926407f2534.png)
As you can see, the furthest point is not visible behind the central point, while the central one is visible behind the closest... Something fishy is up! :)

```python
import numpy as np
import vispy.scene
from vispy.scene import visuals

canvas = vispy.scene.SceneCanvas(keys='interactive', show=True)
view = canvas.central_widget.add_view()

np.random.seed(1)
pos = np.random.normal(size=(3, 3))

scatter = visuals.Markers()
scatter.set_data(pos, edge_color='blue', face_color=(1, 1, 1, 0.5), size=0.5, edge_width=10, scaling=True)

view.add(scatter)

view.camera = 'turntable'
view.camera.set_state({
    'scale_factor': 4.096401677030079,
    'center': [0.33592164516448975, 0.05210036039352417, -0.9912497997283936],
    'fov': 0.0,
    'elevation': 33.5,
    'azimuth': 57.5,
    'roll': 0.0
})

vispy.app.run()
```

**Comment by brisvag** [comment]
Yeah, after some read up, I'm pretty sure this is simply what #2039 was addressing.

**Comment by asnt** [comment]
As @djhoese hinted at, disabling depth testing allows all markers to be drawn.

```diff
--- bug-2208.py	2021-09-01 18:34:13.681330056 +0200
+++ bug-2208-fix.py	2021-09-01 18:40:07.619931629 +0200
@@ -12,6 +12,7 @@
 
 scatter = visuals.Markers()
 scatter.set_data(pos, edge_color='blue', face_color=(1, 1, 1, 0.5), size=0.5, edge_width=10, scaling=True)
+scatter.update_gl_state(depth_test=False)
 
 view.add(scatter)
```

![bug-2208-fix-capture](https://user-images.githubusercontent.com/830589/131710846-95a15e0f-8c25-42a7-9c01-b9e1570e6ef0.png)

**Comment by asnt** [comment]
Right. Then, as you said, an order-independent transparency method like #2039 seems required.

**Comment by brisvag** [comment]
:+1: ! I would love to help out with moving that PR along, if you like (and if there's something easy enought for a gl newbie like me :)). Is there some part of that I could take on?

**Comment by asnt** [comment]
@brisvag I think the general mechanism for rendering transparency is in place, but that some adaptions in the MarkersVisual are needed. It might not be that much, though. Not sure without a test. I would try running an example and continue from there. See also the PR for some details on the limitations/mechanism.

**Comment by asnt** [comment]
(Moreover, a limitation of this OIT algorithm would be that opaque and transparent markers would need to be separated into two different MarkersVisuals (and this is the case for all visual types), when not all markers are transparent. But that is a subsequent consideration, after making it work at all with only transparent objects.)

**Comment by rougier** [comment]
Bit late on this thread but I think the problem you encoutered is actually related to order independent transparency as you suggested. Markers are rendered in random orders and the depth is used to decide if a frag need to be rendered or not. When you have transparency, all the frags are drawn but then the result will vary depending on the the order markers are rendered (if you move the camera, the relative order will change). This order is important because when you blend your color with the background, what you want is blend with what will be present eventually in the scene while what you end up with is a blendinc with what is currently present in the scene and then you get inconsistenies. This is where OIT becoms necessarys. We did not implement it for marker because the transpareny is only used for markers edge but if you look closely at each example, you can see the problem.
